CRM/Civicase/Hook/BuildForm/DisplayAllCustomGroupsInCaseForm.php

```diff
@@ -0,0 +1,56 @@
+<?php
+
+/**
+ * Display all custom groups in case form hook.
+ */
+class CRM_Civicase_Hook_BuildForm_DisplayAllCustomGroupsInCaseForm {
+
+  /**
+   * Adds all custom groups (inline/tabs) to the case form.
+   *
+   * @param CRM_Core_Form $from
+   *   CiviCRM Form
+   */
+  public function run ($form) {
+    if (!$this->shouldRun($form)) {
+      return;
+    }
+
+    $caseCategory = $form->getVar('_type');
+    $caseId = CRM_Utils_Request::retrieve('entityID', 'Positive');
+    $caseType = CRM_Utils_Request::retrieve('subType', 'Positive');
+    $customGroups = CRM_Civicase_APIHelpers_CustomGroups::getAllActiveGroupsForEntity($caseCategory);
```

Comment:
We are fetching all active custom groups for the Case entity here. Not that the create case form also works a bit differently, not sure if you have verified that this works fine. 

When the create case form is viewed without selecting any Case Type, the custom fields that extends all case types is shown as part of the form.

Then when a particular case type is selected, the custom fields that extends that particular case type alone is displayed in addition to custom fields that extends all case types. 

Will be good to check what happens in these scenarios to be sure that this works as expected. e.g if a custom group set is set to be tabbed and set only to extend a particular case type, does this custom field set show even when this case type is not selected?

We should also check if the custom field values are retained when there is validation error on the form, e.g a required field is left empty and the form saved.

Had some issues with these in the past that's why I'm giving a heads up to check.

Reply:
@tunbola thanks for your input. I changed the parameter names because I was not using them right as you mentioned. The case category is never passed to this form, `_type` will always refer to the entity, which is `Case` for us. The way to determine the case category is by using the case type returned by the `subType` parameter. `CRM_Civicase_APIHelpers_CustomGroups::getAllActiveGroupsForEntity` returns ALL custom groups for all case types and categories, but `CRM_Core_BAO_CustomGroup::getTree` will only return the right custom fields for the given case type and other groups are removed from the results array. This is how it works on the original form. Since this was very confusing I added a few notes.

Also, to support validation we need to target the case form so the hook now runs for both forms.